ruby/computer_science/time_complexity.md

```diff
@@ -0,0 +1,306 @@
+
+We can keep doing this until we have an array of just one item. If it matches the number we're looking for we've found it, if not then it isn't in the array.
+
+The below table summarises the size of an array doubling and how many steps in Big O terms we would need to arrive at one element to see if it matches what we're looking for
+
+| Size | Steps |
+| ----------- | ----------- |
+| 1 | 0 |
+| 2 | 1 |
+| 4 | 2 |
+| 8 | 3 |
+| 16 | 4 |
+| 32 | 5 |
+
+Pretty impressive eh!
+
+#### O(N) - Linear Complexity
+
+This one is pretty easy to wrap your head around. Linear Complexity just tells us that as the number of items grows the number of steps grows at exactly the same rate. Every time you iterate over an array is an example of Linear Complexity. If you have an array of 5 items then we can iterate every element in 5 steps. An array of 10 items can be iterated in 10 steps. If you come across any algorithm with a Big O efficiency of `O(N)`, you know that the number of steps will increase in line with the number of elements in your data structure.
+
+#### O(N log N) - N x log N Complexity
+
+You can't say this one isn't appropriately named. This notation means we have an algorithm which initially is `O(log N)` such as our example earlier of Binary Search where it repeatedly breaks an array in half, but with `O(N log N)` each of those array halves is processed by another algorithm with a complexity of `O(N)`
+
+One such algorithm is merge sort and it just so happens you tackle this project in our course :)
+
+#### O(&#178;) - Quadratic Complexity
+
+You've probably written code with a Quadratic Complexity on your programming journey. It's commonly seen when you loop over a data set and within each loop you loop over it again. Therefore when the number of items in the data increases by 1, it requires 2 extra iterations, 2 extra items requires 4 extra iterations (2 in the outer loop and two in the inner loop). 3 extra items adds 9 steps and 4 adds 16 extra steps. We hope you can see where we're going with this...
+
+#### O(&#179;) - Cubic Complexity
+
+Think triple nested loops baby. 1 extra item adds 3 extra steps, 2 adds 8 and 3 adds about 27. 100 items will be about 1,000,000 steps. Ouch!
+
+#### O(2&#8319;) - Exponential Complexity
+
+Exponential Complexity means that with each item added to the data size, the number of steps doubles from the previous number of steps. Let's provide a little table to see how quickly this can get out of hand.
+
+| Size | Steps |
+| ----------- | ----------- |
+| 1 | 1 |
+| 2 | 2 |
+| 3 | 4 |
+| 4 | 8 |
+| 5 | 16 |
+| 6 | 32 |
+| 7 | 64 |
+| 8 | 128 |
+| 9 | 256 |
+| 10 | 512 |
```

Comment:
Still not sure about the math here. For an exponential time complexity, shouldn't we get `2` steps for a data size of `1`?